### PR TITLE
add env addresses to address strategy. Resolves #208

### LINKED ISSUE #208: address `SearchStrategy` should fuzz addresses from the environment
As per title.

This has already been referenced here: https://github.com/vyperlang/titanoboa/blob/master/boa/test/strategies.py#L103.

I think this feature would greatly improve the quality of address fuzzing. From my experience with other frameworks that support this, it always ends up catching edge cases.

### PR DESCRIPTION
**Pull request by theodorecurtil**
### What I did

included in address strategy all env contract addresses + eoa

### CODE REVIEW

File: boa/test/strategies.py [hunk 1]
```diff
@@ -100,11 +101,14 @@ def generate_random_string(n):
...
 def _address_strategy(length: Optional[int] = 100) -> SearchStrategy:
     random_strings = generate_random_string(length)
-    # TODO: add addresses from the environment. probably everything in
-    # boa.env._contracts, boa.env._blueprints and boa.env.eoa.
     accounts = [format_addr(i) for i in random_strings]
+
+    # inject env addresses, ultimately, will sample length + 1 + len(boa.env._contracts) addresses
+    accounts += [boa.env.eoa]
+    accounts += [contract.address for contract in boa.env._contracts.values()]
+    # TODO: add all addresses generated by env through env.generate_address()
```
Comment (DanielSchiavini) [inline boa/test/strategies.py:109]: - I don't think it's very likely, but it seems more correct to put the addresses in a `set`. Otherwise, we don't need to call `list` in line 111 - `accounts` is already a list
- The TODO should be removed
- Why aren't we including blueprints like decribed in the TODO?

```suggestion
    accounts = {format_addr(i) for i in random_strings} & {boa.env.eoa} & \
        {contract.address for contract in boa.env._contracts.values()} & \
        {blueprint.address for blueprint in boa.env._blueprints.values()}
```

Comment (theodorecurtil) [reply boa/test/strategies.py:109]: - The TODO is a new TODO. We discussed with @charles-cooper to maybe buffer all addresses generated by env.generate_address() in a new env attribute and include them in the address strategy. Thoughts?
- re blueprints, i do not see a _blueprints attribute in the env class. Was it replaced by _code_registry?

Comment (DanielSchiavini) [reply boa/test/strategies.py:109]: `env._code_registry` includes both blueprints and contracts
saving all addresses seems like a good idea!

Comment (theodorecurtil) [reply boa/test/strategies.py:109]: shall i make a new issue for saving all generated addresses? Or use this one? @charles-cooper suggested to use [weakref.WeakSet](https://docs.python.org/3/library/weakref.html#weakref.WeakSet) to buffer generated addresses.

something like:

```
def generate_address(self, alias: Optional[str] = None) -> _AddressType:
        t = Address(self._random.randbytes(20))
        if alias is not None:
            self.alias(t, alias)

        # add to weakref.weakset _addresses
        self._addresses.add(t)
        return t
```

Comment (theodorecurtil) [reply boa/test/strategies.py:109]: The strategy would become

```
@_exclude_filter
def _address_strategy(length: Optional[int] = 100) -> SearchStrategy:
    random_strings = generate_random_string(length)

    accounts = ({format_addr(i) for i in random_strings} |
                {boa.env._addresses} |
                {obj.address for obj in boa.env._code_registry.values()})

    return _DeferredStrategyRepr(
        lambda: st.sampled_from(list(accounts)), "accounts"
    )
```

Comment (DanielSchiavini) [reply boa/test/strategies.py:109]: That looks good! Maybe we could generate less random strings depending on how many addresses we have in the env.

Comment (theodorecurtil) [reply boa/test/strategies.py:109]: unfortunately, basic data types cannot be weakly referenced (including str), and Address class extends the str class. Do we need Address to extend str? Is it actually used? Can we add some methods to Address to be directly comparable to str without extending str?

Comment (theodorecurtil) [reply boa/test/strategies.py:109]: maybe we could use a custom overflowing set, but that s not so elegant.

Comment (theodorecurtil) [reply boa/test/strategies.py:109]: (the idea is to avoid having the memory footprint of the env getting out of control as more generated addresses are buffered)

Comment (DanielSchiavini) [reply boa/test/strategies.py:109]: We could use the existing [`lrudict`](https://github.com/vyperlang/titanoboa/blob/1bf16f916b91aab299e293b4148be455c1674706/boa/util/lrudict.py) or a similar structure for sets to keep the memory footprint limited.

The weakset doesn't make sense indeed for simple types as they always get copied and never referenced.

Comment (charles-cooper) [reply boa/test/strategies.py]: > opted in to use lrudict to keep a cache of latest generated addresses, to use in the address strategy

hang on, i think this should be out of scope as it requires a little more thought. can we revert this change? i thought it was fine as an initial change before

Comment (theodorecurtil) [reply boa/test/strategies.py]: ok i ll keep a TODO in the strat then

Comment (charles-cooper) [reply boa/test/strategies.py]: > unfortunately, basic data types cannot be weakly referenced (including str), and Address class extends the str class. Do we need Address to extend str? Is it actually used? Can we add some methods to Address to be directly comparable to str without extending str?

yes, we do, because it compares to string cleanly in user tests. let's revisit it later then

Comment (charles-cooper) [reply boa/test/strategies.py]: > That looks good! Maybe we could generate less random strings depending on how many addresses we have in the env.

no, let's keep generating random strings - we don't want to reduce fuzzing strength just bc a user already has addresses in the env

Comment (charles-cooper) [reply boa/test/strategies.py]: > `env._code_registry` includes both blueprints and contracts saving all addresses seems like a good idea!

i was a little hesitant about this because blueprints can't really do things like make contract calls. but maybe that's actually a good thing? because then the user has to deal with addresses which contain un-executable code.

Comment (theodorecurtil) [reply boa/test/strategies.py]: btw, i saw that the random addresses are all like "0x0000000000000000000whatever", not covering the complete "bytespace" at all

File: boa/test/strategies.py [hunk 2]
```diff
@@ -99,12 +100,18 @@ def generate_random_string(n):
 
 @_exclude_filter
 def _address_strategy(length: Optional[int] = 100) -> SearchStrategy:
+
+    # TODO: add addresses generated by the env through the env.generate_address() method
+    env_accounts = (
+            {obj.address for obj in boa.env._code_registry.values()}
```
Comment (charles-cooper) [inline boa/test/strategies.py:106]: it's a bit weird here to introduce a dependency on the global env. should we maybe make it parametrizable? like if `None` is provided then we use `boa.env`, but otherwise a specific env can be given.

File: boa/test/strategies.py (hunk 1 above)
Comment (charles-cooper) [reply boa/test/strategies.py]: > btw, i saw that the random addresses are all like "0x0000000000000000000whatever", not covering the complete "bytespace" at all

yea that's busted, it's the existing definition for `generate_random_string`:
https://github.com/theodorecurtil/titanoboa/blob/1b54ccfc3f5690ee4b6cb6f6a1cfbd94cad47188/boa/test/strategies.py#L98C21-L98C35

honestly we should probably switch to something like https://docs.python.org/3/library/random.html#random.randbytes

File: boa/test/strategies.py (hunk 2 above)
Comment (theodorecurtil) [reply boa/test/strategies.py:106]: something like that?
```
@_exclude_filter
def _address_strategy(length: Optional[int] = 100, environment: Optional[EnvType] = None) -> SearchStrategy:

    if environment is None:
        environment = boa.env

    # TODO: add addresses generated by the env through the env.generate_address() method
    env_accounts = (
            {obj.address for obj in environment._code_registry.values()}
    )

    # generate `length` random addresses
    random_strings = generate_random_string(length)
    random_accounts = {format_addr(i) for i in random_strings}

    return _DeferredStrategyRepr(
        lambda: st.sampled_from(list(env_accounts | random_accounts)), "accounts"
    )
```

Then would need to modify all tests to pass a specific env

Comment (theodorecurtil) [reply boa/test/strategies.py:106]: strategies are only used by pytest right? it would make sense to use the environment that is available in the context of the tests.

Comment (charles-cooper) [reply boa/test/strategies.py:106]: that's true, but you can set up contracts without using the global env

File: boa/test/strategies.py [hunk 3]
```diff
@@ -98,13 +101,22 @@ def generate_random_string(n):
...
+    # generate `length` random addresses
     random_strings = generate_random_string(length)
-    # TODO: add addresses from the environment. probably everything in
-    # boa.env._contracts, boa.env._blueprints and boa.env.eoa.
-    accounts = [format_addr(i) for i in random_strings]
+    random_accounts = {format_addr(i) for i in random_strings}
+
     return _DeferredStrategyRepr(
-        lambda: st.sampled_from(list(accounts)[:length]), "accounts"
+        lambda: st.sampled_from(list(env_accounts | random_accounts)), "accounts"
```
Comment (charles-cooper) [inline boa/test/strategies.py:119]: i think using `st.one_of(st.sampled_from(list(env_accounts), list(random_accounts)))` will generate a slightly better distribution

we can also use `st.binary()` for the random addresses actually
